File: Project/line.js

const Matrix3D = require('./matrix');
const Quaternion = require('./quaternion');
const Vector3D = require('./vectors');

class Line3D{
    constructor(p, dir){
        this.p = p;
        this.dir = dir;
    }

    linesIntersection(line2){
        const crossProduct = this.dir.crossProduct(line2.dir);

        const epsilon = 1e-10;

        if (Math.abs(crossProduct.calculateLength()) < epsilon) {
            console.log("Lines are parallel or coincident.");
            return null;
        }
        //punkt przecięcia na prostej (długość jakby przez to mnożymy)
        const t = line2.dir.crossProduct(this.p.subtract(line2.p)).dotProduct(crossProduct) / crossProduct.calculateLength();

        const intersectionp = this.p.add(this.dir.scalarMultiply(t));

        return intersectionp;
    }
    
    AngleBetweenLines(otherLine) {
        const dotProduct = this.dir.dotProduct(otherLine.dir);
        const magnitudeProduct = this.dir.calculateLength() * otherLine.dir.calculateLength();

        if (magnitudeProduct === 0) {
            console.log("One of the lines has zero magnitude.");
            return null;
        }

        const cosineTheta = dotProduct / magnitudeProduct;
        const angleRad = Math.acos(cosineTheta);
        const angleDegrees = (180 / Math.PI) * angleRad;

        return angleDegrees;
    }

    intersectionWithPlane(plane) {
        const ts = plane.a * this.dir.x + plane.b * this.dir.y + plane.c * this.dir.z;
        const values = -plane.d - (plane.a * this.p.x + plane.b * this.p.y + plane.c * this.p.z);
        const t = values / ts;

        return new Vector3D(
            this.p.x + this.dir.x * t,
            this.p.y + this.dir.y * t,
            this.p.z + this.dir.z * t
        );
    }

    anglePlane(plane) {
        const v4 = new Vector3D(plane.normal.x, plane.normal.y, plane.normal.z).normalize();
        const v5 = new Vector3D(this.dir.x, this.dir.y, this.dir.z).normalize();

        const dotProduct = v4.dotProduct(v5);
        const angleRad = Math.acos(dotProduct);
        const angle = angleRad * (180 / Math.PI);
        const result = 90 - angle;

        return result;
    }
    
}
module.exports = Line3D;

File: Zad6Cpp/src/Line.cpp
#include "../lib/Line.h"

Line::Line()
{
}

Line::Line(const Vector& point, const Vector& direction)
{
	this->point = point;
	this->direction = direction;
}

Vector Line::IntersectionWithLine(Line line)
{
	Vector delta = line.point;
	delta = delta - point;
	Vector cross = direction.cross(line.direction);
	if (std::abs(delta.dotProduct(cross)) > 0.001) {
		return { 0,0,0 };
	}

	float m2 = direction.cross(line.direction).length();
	m2 *= m2;
	Vector cross1 = delta.cross(line.direction);
	Vector cross2 = direction.cross(line.direction);
	float s = cross1.dotProduct(cross2) / m2;

	Vector result = point;
	Vector multiplied = direction;
	multiplied.mul(s);
	result = result + multiplied;
	return result;

}

Vector Line::IntersectionWithPlane(Plane plane)
{
	float ts = plane.a * (direction.x) + plane.b * (direction.y) + plane.c * (direction.z);
	float values = -plane.d - (plane.a * (point.x) + plane.b * (point.y) + plane.c * (point.z));
	float t = values / ts;

	return Vector(point.x + direction.x * t, point.y + direction.y * t, point.z + direction.z * t);
}

float Line::AngleLine(Line line)
{
	Vector v4(direction.x, direction.y, direction.z);
	Vector v5(line.direction.x, line.direction.y, line.direction.z);

    v4.normalize();
    v5.normalize();

	float dotProduct = v4.dotProduct(v5);

    float angleRad = acos(dotProduct);
    float angle = angleRad * 180 / M_PI;

    return angle;
}

float Line::AnglePlane(Plane plane)
{
	Vector v4(plane.normal.x, plane.normal.y, plane.normal.z);
	Vector v5(direction.x, direction.y, direction.z);

	v4.normalize();
    v5.normalize();

    float dotProduct = v4.dotProduct(v5);

    float angleRad = acos(dotProduct);
    float angle = angleRad * 180 / M_PI;
    float result = 90 - angle;

    return result;
}

std::ostream& operator<<(std::ostream& strm, const Line& line) {

	return strm << line.point << " " << line.direction;
}